tentacle/src/transports/tls.rs

```diff
@@ -0,0 +1,275 @@
+use super::Result;
+use futures::{future::ok, FutureExt, SinkExt, Stream, TryFutureExt};
+use log::warn;
+use std::{
+    borrow::Cow,
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+use crate::runtime::TcpListener;
+use crate::service::TlsConfig;
+use crate::{
+    error::TransportErrorKind,
+    multiaddr::{Multiaddr, Protocol},
+    session::AsyncRw,
+    transports::{tcp_dial, tcp_listen, Transport, TransportFuture},
+    utils::{dns::DnsResolver, multiaddr_to_socketaddr, socketaddr_to_multiaddr},
+};
+use futures::channel::mpsc::{channel, Receiver, Sender};
+use std::sync::Arc;
+use tokio::io;
+use tokio_rustls::rustls::ServerConfig;
+use tokio_rustls::webpki::DNSNameRef;
+use tokio_rustls::{TlsAcceptor, TlsConnector};
+
+pub type TlsStream = Box<dyn AsyncRw + Send + Unpin + 'static>;
+
+/// Tls listen bind
+async fn bind(
+    address: impl Future<Output = Result<Multiaddr>>,
+    timeout: Duration,
+    config: TlsConfig,
+) -> Result<(Multiaddr, TlsListener)> {
+    let addr = address.await?;
+    match multiaddr_to_socketaddr(&addr) {
+        Some(socket_address) => {
+            let (local_addr, tcp) = tcp_listen(socket_address, config.tls_bind.is_some()).await?;
+            let tls_server_config = match config.tls_server_config {
+                Some(tls_server_config) => tls_server_config,
+                None => {
+                    return Err(TransportErrorKind::TlsError(
+                        "server config not found".to_string(),
+                    ));
+                }
+            };
+            let mut listen_addr = socketaddr_to_multiaddr(local_addr);
+            if let Some(domain_name) = parse_tls_domain_name(&addr) {
+                listen_addr.push(Protocol::Tls(Cow::Owned(domain_name)));
+                Ok((
+                    listen_addr,
+                    TlsListener::new(timeout, tcp, tls_server_config),
+                ))
+            } else {
+                Err(TransportErrorKind::NotSupported(addr))
+            }
+        }
+        None => Err(TransportErrorKind::NotSupported(addr)),
+    }
+}
+
+/// Tls connect
+async fn connect(
+    address: impl Future<Output = Result<Multiaddr>>,
+    timeout: Duration,
+    original: Option<Multiaddr>,
+    config: TlsConfig,
+) -> Result<(Multiaddr, TlsStream)> {
+    let tls_client_config = match config.tls_client_config {
+        Some(tls_client_config) => tls_client_config,
+        None => {
+            return Err(TransportErrorKind::TlsError(
+                "client config not found".to_string(),
+            ));
+        }
+    };
+
+    let addr = address.await?;
+    match multiaddr_to_socketaddr(&addr) {
+        Some(socket_address) => {
+            let stream = tcp_dial(socket_address, config.tls_bind, timeout).await?;
+
+            if let Some(domain_name) =
+                parse_tls_domain_name(&original.clone().unwrap_or_else(|| addr.clone()))
+            {
```

Comment:
```suggestion
            let original = original.unwrap_or_else(|| addr);
            if let Some(domain_name) =
                parse_tls_domain_name(&original)
            {
            ...
            Ok((
                    original,
            ...
            } else {
                Err(TransportErrorKind::NotSupported(original))
            }  
```

Reply:
`parse_tls_domain_name` have be done in upper fn & modify `connect` & `bind` fn args